Warp with small objects

### Environment info
- NooBaa Version: **VERSION**
- Platform: **Kubernetes 1.14.1 | minikube 1.1.1 | OpenShift 4.1 | other: specify**
[root@c83f2-dan9 ~]# rpm -qa | grep  s3
gpfs.mms3-5.2.1-0.240617.104624.el9.x86_64
[root@c83f2-dan9 ~]# rpm -qa | grep  noobaa
noobaa-core-5.15.4-20240611.el9.x86_64



### Actual behavior
1. when im trying to run 100 warp instances (25 for each app node) in order to use 100 accounts (one for each instance) in addition to using 20 connections per instance (2000 in total) for small objects where these errors were seen:


noobaa.log:

2024-06-07T06:13:29.739738-04:00 c83f2-dan9 node[3192156]: [nsfs/3192156] [ERROR] CONSOLE:: nsfs: exit on error Error: failed to create NSFS system data due to - ENOENT: no such file or directory, unlink '/gpfs/remote_rw_cessharedroot/ces/s3-config/system.json.lx4j3mkp-458a72e6'    at init_nsfs_system (/usr/local/noobaa-core/src/cmd/nsfs.js:233:23)    at async main (/usr/local/noobaa-core/src/cmd/nsfs.js:320:27)
2024-06-07T06:13:34.236337-04:00 c83f2-dan9 node[3192178]: [nsfs/3192178] [ERROR] CONSOLE:: failed to create NSFS system data due to - ENOENT: no such file or directory, unlink '/gpfs/remote_rw_cessharedroot/ces/s3-config/system.json.lx4j3q1h-fbdf2057' [Error: ENOENT: no such file or directory, unlink '/gpfs/remote_rw_cessharedroot/ces/s3-config/system.json.lx4j3q1h-fbdf2057'] { errno: -2, code: 'ENOENT', syscall: 'unlink', path: '/gpfs/remote_rw_cessharedroot/ces/s3-config/system.json.lx4j3q1h-fbdf2057' }


### Expected behavior
1. 100 instances of warp complete

### Steps to reproduce
1. run 100 instances of warp using 20 conecctions per node

### More information - Screenshots / Logs / Other output

--- a/src/agent/block_store_services/block_store_fs.js
+++ b/src/agent/block_store_services/block_store_fs.js
@@ -43,7 +43,7 @@ class BlockStoreFs extends BlockStoreBase {
         this.xattr_enabled = true;
 
         this.fs_context = {
-            disable_ctime_check: config.BLOCK_STORE_FS_TMFS_ENABLED
+            do_ctime_check: !config.BLOCK_STORE_FS_TMFS_ENABLED
         };
     }
 

--- a/src/native/fs/fs_napi.cpp
+++ b/src/native/fs/fs_napi.cpp
@@ -559,9 +559,9 @@ struct FSWorker : public Napi::AsyncWorker
     double _took_time;
     Napi::FunctionReference _report_fs_stats;
 
-    // disables the ctime check in the stat and read file fuctions
-    // NOTE: If this is enabled then some functions will fallback to using mtime check
-    bool _disable_ctime_check;
+    // executes the ctime check in the stat and read file fuctions
+    // NOTE: If _do_ctime_check = false, then some functions will fallback to using mtime check
+    bool _do_ctime_check;
 
     FSWorker(const Napi::CallbackInfo& info)
         : AsyncWorker(info.Env())
@@ -573,7 +573,7 @@ struct FSWorker : public Napi::AsyncWorker
         , _errno(0)
         , _warn_threshold_ms(0)
         , _took_time(0)
-        , _disable_ctime_check(false)
+        , _do_ctime_check(false)
     {
         for (int i = 0; i < (int)info.Length(); ++i) _args_ref.Set(i, info[i]);
         if (info[0].ToBoolean()) {
@@ -589,7 +589,7 @@ struct FSWorker : public Napi::AsyncWorker
             if (fs_context.Get("report_fs_stats").ToBoolean()) {
                 _report_fs_stats = Napi::Persistent(fs_context.Get("report_fs_stats").As<Napi::Function>());
             }
-            _disable_ctime_check = fs_context.Get("disable_ctime_check").ToBoolean();
+            _do_ctime_check = fs_context.Get("do_ctime_check").ToBoolean();
         }
     }
     void Begin(std::string desc)
@@ -744,7 +744,7 @@ struct Stat : public FSWorker
             }
         }
 
-        if (!_disable_ctime_check) {
+        if (_do_ctime_check) {
             CHECK_CTIME_CHANGE(fd, _stat_res, _path);
         }
     }
@@ -1191,10 +1191,10 @@ struct Readfile : public FSWorker
             p += len;
         }
 
-        if (_disable_ctime_check) {
-            CHECK_MTIME_CHANGE(fd, _stat_res, _path);
-        } else {
+        if (_do_ctime_check) {
             CHECK_CTIME_CHANGE(fd, _stat_res, _path);
+        } else {
+            CHECK_MTIME_CHANGE(fd, _stat_res, _path);
         }
     }
     virtual void OnOK()
@@ -1689,7 +1689,7 @@ struct FileStat : public FSWrapWorker<FileWrap>
             GPFS_FCNTL_OR_RETURN(get_fd_gpfs_xattr(fd, _xattr, gpfs_error));
         }
 
-        if (!_disable_ctime_check) {
+        if (_do_ctime_check) {
             CHECK_CTIME_CHANGE(fd, _stat_res, _wrap->_path);
         }
     }

--- a/src/sdk/namespace_fs.js
+++ b/src/sdk/namespace_fs.js
@@ -1267,8 +1267,8 @@ class NamespaceFS {
         const part_upload = file_path === upload_path;
         const same_inode = params.copy_source && copy_res === copy_status_enum.SAME_INODE;
         const is_dir_content = this._is_directory_content(file_path, params.key);
-        // upload_part should disable ctime_check because we update the same part-file on concurrent put part 
-        let stat = await target_file.stat({ ...fs_context, disable_ctime_check: part_upload });
+
+        let stat = await target_file.stat(fs_context);
         this._verify_encryption(params.encryption, this._get_encryption_info(stat));
 
         // handle xattr
@@ -1320,7 +1320,7 @@ class NamespaceFS {
             if (params.copy_source) fs_xattr = await this._get_copy_source_xattr(params, fs_context, fs_xattr);
             await this._assign_dir_content_to_xattr(fs_context, fs_xattr, { ...params, size: stat.size }, copy_xattr);
         }
-        stat = await nb_native().fs.stat({ ...fs_context, disable_ctime_check: part_upload }, file_path);
+        stat = await nb_native().fs.stat(fs_context, file_path);
         const upload_info = this._get_upload_info(stat, fs_xattr && fs_xattr[XATTR_VERSION_ID]);
         return upload_info;
     }
@@ -2419,7 +2419,7 @@ class NamespaceFS {
         await this._load_bucket(params, fs_context);
         params.mpu_path = this._mpu_path(params);
         try {
-            await nb_native().fs.stat({ ...fs_context, disable_ctime_check: true }, params.mpu_path);
+            await nb_native().fs.stat(fs_context, params.mpu_path);
         } catch (err) {
             // TOOD: Error handling
             if (err.code === 'ENOENT') err.rpc_code = 'NO_SUCH_UPLOAD';

--- a/src/sdk/nb.d.ts
+++ b/src/sdk/nb.d.ts
@@ -1020,7 +1020,7 @@ interface NativeFSContext {
     backend?: string;
     warn_threshold_ms?: number;
     report_fs_stats?: Function;
-    disable_ctime_check?: boolean;
+    do_ctime_check?: boolean;
 }
 
 type GPFSNooBaaArgs = {

